bin/zowe-setup-certificates.sh

```diff
@@ -0,0 +1,176 @@
+    ${ZOWE_ROOT_DIR}/bin/apiml_cm.sh --verbose --log $LOG_FILE --action setup --service-ext ${SAN} --service-password ${KEYSTORE_PASSWORD} \
+      --service-alias ${KEYSTORE_ALIAS} --service-keystore ${KEYSTORE_PREFIX} --service-truststore ${TRUSTSTORE_PREFIX} --local-ca-filename ${LOCAL_CA_PREFIX}
+    RC=$?
+    echo "apiml_cm.sh --action setup returned: $RC" >> $LOG_FILE
+  else
+    (>&2 echo "Zowe Install setup configuration is invalid; check your zowe-setup-certificates.env file.")
+    (>&2 echo "Some external apiml certificate fields are supplied...Fields must be filled out in full or left completely blank.")
+    (>&2 echo "See $LOG_FILE for more details.")
+    echo "</zowe-setup-certificates.sh>" >> $LOG_FILE
+    exit 1
+  fi
+else
+  EXT_CA_PARM=""
+  for CA in ${EXTERNAL_CERTIFICATE_AUTHORITIES}; do
+      EXT_CA_PARM="${EXT_CA_PARM} --external-ca ${CA} "
+  done
+
+  ${ZOWE_ROOT_DIR}/bin/apiml_cm.sh --verbose --log $LOG_FILE --action setup --service-ext ${SAN} --service-password ${KEYSTORE_PASSWORD} \
+    --external-certificate ${EXTERNAL_CERTIFICATE} --external-certificate-alias ${EXTERNAL_CERTIFICATE_ALIAS} ${EXT_CA_PARM} \
+    --service-alias ${KEYSTORE_ALIAS} --service-keystore ${KEYSTORE_PREFIX} --service-truststore ${TRUSTSTORE_PREFIX} --local-ca-filename ${LOCAL_CA_PREFIX} \
+    --external-ca-filename ${EXTERNAL_CA_PREFIX}
+  RC=$?
+
+  echo "apiml_cm.sh --action setup returned: $RC" >> $LOG_FILE
+fi
+
+if [ "$RC" -ne "0" ]; then
+    (>&2 echo "apiml_cm.sh --action setup has failed. See $LOG_FILE for more details")
+    echo "</zowe-setup-certificates.sh>" >> $LOG_FILE
+    exit 1
+fi
+
+if [[ "${VERIFY_CERTIFICATES}" == "true" ]]; then
+  ${ZOWE_ROOT_DIR}/bin/apiml_cm.sh --verbose --log $LOG_FILE --action trust-zosmf --zosmf-keyring ${ZOSMF_KEYRING} --zosmf-userid ${ZOSMF_USER} \
+    --service-password ${KEYSTORE_PASSWORD} --service-truststore ${TRUSTSTORE_PREFIX}
+  RC=$?
+
+  echo "apiml_cm.sh --action trust-zosmf returned: $RC" >> $LOG_FILE
+  if [ "$RC" -ne "0" ]; then
+      (>&2 echo "apiml_cm.sh --action trust-zosmf has failed. See $LOG_FILE for more details")
+      (>&2 echo "WARNING: z/OSMF is not trusted by the API Mediation Layer. Follow instructions in Zowe documentation about manual steps to trust z/OSMF")
+      (>&2 echo "  Issue the following command as a user that has permissions to export public certificates from z/OSMF keyring:")
+      (>&2 echo "    ${ZOWE_ROOT_DIR}/bin/apiml_cm.sh --action trust-zosmf --zosmf-keyring ${ZOSMF_KEYRING} --zosmf-userid ${ZOSMF_USER} \
+      --service-password ${KEYSTORE_PASSWORD} --service-truststore ${TRUSTSTORE_PREFIX}")
+  fi
+fi
+
+echo "Creating certificates and keystores... DONE"
+# set up chmod
+chmod -R 600 ${KEYSTORE_DIRECTORY}/${LOCAL_KEYSTORE_SUBDIR}/* ${KEYSTORE_DIRECTORY}/${KEYSTORE_ALIAS}/*
+chown -R ${ZOWE_USER_ID} ${KEYSTORE_DIRECTORY}/${LOCAL_KEYSTORE_SUBDIR} ${KEYSTORE_DIRECTORY}/${KEYSTORE_ALIAS}
```

Comment:
When I ran this I got ```
chown: FSUM6180 file "/u/stevenh/887/keystore/local_ca": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/local_ca/localca.keystore.p12": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/local_ca/localca.cer": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/local_ca/localca.cer-ebcdic": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.keystore.p12": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.keystore.csr": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.keystore_signed.cer": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.truststore.p12": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.keystore.cer": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.keystore.cer-ebcdic": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.keystore.key": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/887/keystore/localhost/localhost.keystore.jwtsecret.pem": EDC5139I Operation not permitted.
```
In the past we've not been able to chown and had to chgrp instead, so that might be worth trying?

Reply:
The current functionality seems to work correctly, but during install I'm still seeing:
```
Creating certificates and keystores... STARTED
Creating certificates and keystores... DONE
Trying to change an owner of the /u/stevenh/zowe/keystore.
chown: FSUM6180 file "/u/stevenh/zowe/keystore": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/local_ca": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/local_ca/localca.keystore.p12": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/local_ca/localca.cer": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/local_ca/localca.cer-ebcdic": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.keystore.p12": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.keystore.csr": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.keystore_signed.cer": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.truststore.p12": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.keystore.cer": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.keystore.cer-ebcdic": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.keystore.key": EDC5139I Operation not permitted.
chown: FSUM6180 file "/u/stevenh/zowe/keystore/localhost/localhost.keystore.jwtsecret.pem": EDC5139I Operation not permitted.
Unable to change the current owner of the /u/stevenh/zowe/keystore directory to the IZUSVR owner.
Trying to change a group of the /u/stevenh/zowe/keystore.
Group of the /u/stevenh/zowe/keystore changed successfully to the IZUADMIN owner.
```
so I'm concerned that customers are going to get worried about the errors. Could we try piping the errors to /dev/null or something, so only the 'managed' printed messages appear?

Reply:
I'll route the error messages to the log file. They could be useful for debugging as there can be more reasons for chown/chgrp to fail.

Reply:
Looking good